src/Deque.mo

```diff
@@ -33,46 +75,158 @@ module {
     }
   };
 
-  /// Insert a new element on the front end of the queue
-  public func pushFront<T>(q : Deque<T>, x : T) : Deque<T> {
-    check(List.push(x, q.0), q.1)
+  /// Extend a deque by inserting a new element on the front end.
+  /// Returns a new deque with `element` in the front followed by the elements of `deque`.
+  ///
+  /// This may involve dynamic splitting of the internally used two lists.
+  ///
+  /// Example:
+  /// ```motoko
+  /// import Deque "mo:base/Deque";
+  ///
+  /// Deque.pushFront(Deque.pushFront(Deque.empty<Nat>(), 2), 1) // deque with elements [1, 2]
+  /// ```
+  ///
+  /// Runtime: `O(n)` worst-case, amortized to `O(1)`.
+  ///
+  /// Space: `O(n)` worst-case, amortized to `O(1)`.
+  ///
+  /// `n` denotes the number of elements stored in the deque.
+  public func pushFront<T>(deque : Deque<T>, element : T) : Deque<T> {
+    check(List.push(element, deque.0), deque.1)
   };
 
-  /// Inspect the (optional) first element on the front end of the queue
-  public func peekFront<T>(q : Deque<T>) : ?T {
-    switch q {
+  /// Inspect the optional element on the front end of a deque.
+  /// Returns `null` if `deque` is empty. Otherwise, the front element of `deque`.
+  ///
+  /// Example:
+  /// ```motoko
+  /// import Deque "mo:base/Deque";
+  ///
+  /// let deque = Deque.pushFront(Deque.pushFront(Deque.empty<Nat>(), 2), 1);
+  /// Deque.peekFront(deque) // => ?1
+  /// ```
+  ///
+  /// Runtime: `O(1)`.
+  ///
+  /// Space: `O(1)`.
+  ///
+  public func peekFront<T>(deque : Deque<T>) : ?T {
+    switch deque {
       case (?(x, f), r) { ?x };
       case (null, ?(x, r)) { ?x };
       case _ { null }
     }
   };
 
-  /// Remove the first element on the front end of the queue; Returns null when empty.
-  public func popFront<T>(q : Deque<T>) : ?(T, Deque<T>) {
-    switch q {
+  /// Shorten a deque by removing the element on the front end.
+  /// Returns `null` if `deque` is empty. Otherwise, it returns a pair of
+  /// the first element and a new deque that contains all the elements of `deque`,
+  /// however, without the front element.
+  ///
+  /// This may involve dynamic splitting of the internally used two lists.
+  ///
+  /// Example:
+  /// ```motoko
+  /// import Deque "mo:base/Deque";
+  ///
+  /// let initial = Deque.pushFront(Deque.pushFront(Deque.empty<Nat>(), 2), 1);
+  /// let reduced = Deque.popFront(initial);
+  /// switch reduced {
+  ///     case null assert(false);
```

Comment:
So I think new users might have a hard time realizing that `assert false` corresponds to an impossible case on first glance. 

It might be nicer to have `{ Debug.trap "Empty queue impossible" };`. It seems a bit more clear.

Also, since trap has type `None`, the interpreter won't complain that there are mismatching types between the two branches (`Nat` and `()`).

Reply:
Thanks for this very useful suggestion. I improved this.